File: nvpl_blas/f77/zhpr.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     ZHPR 
!
!******************************************************************************
      program   ZHPR_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      integer         n, incx, i
      parameter       (n=2, incx=1)
      double precision alpha
      parameter       (alpha=1.0)
      integer         xsize
      parameter       (xsize=1+(n-1)*abs(incx))
      double complex  x(xsize), a(n,n)

*       External Subroutines
      external        print_zvector, fill_zvector, fill_zmatrix,
     &                print_zmatrix
      character*1     uplo
      parameter       (uplo='U')
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx
      print 101, alpha
      call fill_zvector(x,n,incx)
      call fill_zmatrix(a,n,n,n,uplo,'U')
      call print_zmatrix(a,n,n,n,'A')
      call print_zvector(x,n,incx,'X')
*
*      Call ZHPR subroutine
      call ZHPR(uplo,n,alpha,x,incx,a)
*
      print*
      print 102
      call print_zmatrix(a,n,n,n,'A')

      stop
 99   format('Example: ZHPR for the packed Hermitian rank-1 update')
 100  format('#### args: n=',i1,', incx=',i1)
 101  format(11x,'alpha=',f4.2)
 102  format('The result of ZHPR: ')
      end

File: nvpl_blas/c/zhpr.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_zhpr
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t incX = 1;
    double alpha = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    nvpl_dcomplex_t * A = NULL;
    nvpl_dcomplex_t * X = NULL;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);

    printf("\nExample: cblas_zhpr for the Hermitian rank-1 update\n\n");
    printf("#### args: n=%" PRId64 ", incX=%" PRId64 ", order=%c, uplo=%c\n",
           (int64_t)N, (int64_t)incX, order_to_char(order), uplo_to_char(uplo));
    printf("           alpha=%g\n", alpha);

    // allocate memory
    A = (nvpl_dcomplex_t *)malloc(N * N  * sizeof(nvpl_dcomplex_t));
    X = (nvpl_dcomplex_t *)malloc(len_x * sizeof(nvpl_dcomplex_t));

    // fill data
    fill_zmatrix(A, N, N, N, order, UPLO_TO_FILL_MODE(uplo), CblasUnit);
    fill_zvector(X, N, incX);

    // print input data
    print_zmatrix(A, N, N, N, "A", order);
    print_zvector(X, N, incX, "X");

    // call cblas_zhpr
    cblas_zhpr(order, uplo, N, alpha, X, incX, A);

    // print result
    printf("\nAfter zhpr operation:\n");
    print_zmatrix(A, N, N, N, "A", order);

    // release memory
    free(A);
    free(X);
    return 0;
}
